### PR TITLE
Fix array equality check while publishing properties

### PR DESCRIPTION
**Pull request by nedimtokic**
In C# "Equals" method compare arrays by reference. Instead of it, when the payload is "Array" we compare arrays by each element.

### CODE REVIEW

File: AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs [hunk 1]
```diff
@@ -54,6 +54,50 @@ public void SetTimestamp(DateTime timestamp)
...
+            if (otherPayload is null)
+            {
+                return false;
+            }
+
+            if (payload is Array array && otherPayload is Array otherArray)
+            {
+                if (array == null || otherArray == null) return false;
+                if (array.Length != otherArray.Length) return false;
+                return ArrayEquality(array, otherArray);
```
Comment (harlem88) [inline AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs:69]: What do you think about using [Enumerable.SequenceEqual](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.sequenceequal?redirectedfrom=MSDN&view=net-7.0#System_Linq_Enumerable_SequenceEqual__1_System_Collections_Generic_IEnumerable___0__System_Collections_Generic_IEnumerable___0__) ? 
```suggestion
                return Enumerable.SequenceEqual(array, otherArray);
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs (hunk 1 above)
Comment (nedimtokic) [reply AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs:69]: We don't know TSource at this point. We must cast both arrays to their data type to implement this method. 
(etc. `return Enumerable.SequenceEqual((IEnumerable<int>)array, (IEnumerable<int>)otherArray)`)

File: AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs [hunk 2]
```diff
@@ -54,6 +54,50 @@ public void SetTimestamp(DateTime timestamp)
...
+                        return false;
+                    }
+                }
+                else
+                {
+                    if (!payload.GetValue(i)!.GetHashCode().Equals(otherArray.GetValue(i)!.GetHashCode()))
+                    {
+                        return false;
+                    }
+                }
```
Comment (harlem88) [inline AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs:96]: Use `else if` block
```suggestion
                else if (!payload.GetValue(i)!.GetHashCode().Equals(otherArray.GetValue(i)!.GetHashCode()))
                {                   
                      return false;                    
                }
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs (hunk 2 above)
Comment (nedimtokic) [reply AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs]: Changed

File: AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs [hunk 3]
```diff
@@ -54,6 +54,47 @@ public void SetTimestamp(DateTime timestamp)
...
+                if (array == null || otherArray == null) return false;
+                if (array.Length != otherArray.Length) return false;
+                return ArrayEquality(array, otherArray);
+            }
+
+            return payload.GetHashCode().Equals(otherPayload.GetHashCode());
+
+        }
+
+        private bool ArrayEquality(Array payload, Array otherArray)
```
Comment (harlem88) [inline AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs:76]:
```suggestion
        private bool ArrayEquality(Array array, Array otherArray)
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs (hunk 3 above)
Comment (nedimtokic) [reply AstarteDeviceSDKCSharp/Utilities/DecodedMessage.cs]: Local variable renamed to 'array'.

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 - Fix payload validation on the aggregated object interface.
 - Resolve the type inconsistency issue when deserializing BSON to a double array.
+- Resolve equality issue when setting the device property.
```
Comment (harlem88) [inline CHANGELOG.md:11]:
```suggestion
- Resolve array equality issue when setting the device property.
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (nedimtokic) [reply CHANGELOG.md]: Changed.

**Review by osmanhadzic — APPROVED** [review]

**Pull request merged**